OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs

```diff
@@ -10,49 +10,53 @@
 #endregion
 
 using System.Collections.Generic;
+using System.Linq;
 using OpenRA.Mods.Common.Traits;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Warheads
 {
 	public class TargetDamageWarhead : DamageWarhead
 	{
-		public override void DoImpact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
-		{
-			// Damages a single actor, rather than a position. Only support by InstantHit for now.
-			// TODO: Add support for 'area of damage'
-			if (target.Type == TargetType.Actor)
-				DoImpact(target.Actor, firedBy, damageModifiers);
-		}
+		[Desc("Damage will be applied to actors in this area. A value of zero means only targeted actor will be damaged.")]
+		public readonly WDist Spread = WDist.Zero;
 
 		public override void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageModifiers)
 		{
-			// For now this only displays debug overlay
-			// TODO: Add support for 'area of effect' / multiple targets
-			var world = firedBy.World;
-			var debugOverlayRange = new[] { WDist.Zero, new WDist(128) };
-
-			var debugVis = world.WorldActor.TraitOrDefault<DebugVisualizations>();
-			if (debugVis != null && debugVis.CombatGeometry)
-				world.WorldActor.Trait<WarheadDebugOverlay>().AddImpact(pos, debugOverlayRange, DebugOverlayColor);
-		}
-
-		public override void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damageModifiers)
-		{
-			if (!IsValidAgainst(victim, firedBy))
+			if (Spread == WDist.Zero)
 				return;
 
-			var damage = Util.ApplyPercentageModifiers(Damage, damageModifiers.Append(DamageVersus(victim)));
-			victim.InflictDamage(firedBy, new Damage(damage, DamageTypes));
-
 			var world = firedBy.World;
+
 			if (world.LocalPlayer != null)
 			{
-				var debugOverlayRange = new[] { WDist.Zero, new WDist(128) };
-
+				var debugOverlayRange = new[] { WDist.Zero, Spread };
 				var debugVis = world.WorldActor.TraitOrDefault<DebugVisualizations>();
 				if (debugVis != null && debugVis.CombatGeometry)
-					world.WorldActor.Trait<WarheadDebugOverlay>().AddImpact(victim.CenterPosition, debugOverlayRange, DebugOverlayColor);
+					world.WorldActor.Trait<WarheadDebugOverlay>().AddImpact(pos, debugOverlayRange, DebugOverlayColor);
+			}
+
+			var hitActors = world.FindActorsOnCircle(pos, Spread);
+
+			foreach (var victim in hitActors)
+			{
+				if (!IsValidAgainst(victim, firedBy))
+					continue;
+
+				var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
+					.MinByOrDefault(t => t.Info.Type.DistanceFromEdge(pos, victim));
+
+				// Cannot be damaged without an active HitShape
+				if (closestActiveShape == null)
+					continue;
+
+				// Fully damage actors whose HitShape intersects with warhead Spread, otherwise don't damage at all
+				var distance = closestActiveShape.Info.Type.DistanceFromEdge(pos, victim);
```

Comment:
This doesn't account for the spread at all!

I'll push a fix to this alongside a few other cleanups.